doc/website/blog/2022-03-14-ent-hooks-examples.md

```diff
@@ -0,0 +1,434 @@
+		if len(sp) != 2 {
+			return next.Mutate(ctx, m)
+		}
+		sp = strings.Split(sp[0], ":")
+		if len(sp) != 3 {
+			return next.Mutate(ctx, m)
+		}
+		pass := sp[2]
+		m.SetPassword(pass)
+		m.SetConnectionString(strings.ReplaceAll(cs, pass, "****"))
+		return next.Mutate(ctx, m)
+	})
+} 
+```
+We validate that the mutation has a connection string (if not just let the mutation continue).
+A bit of string acrobatics, and we have the password extracted from the string, we store it in another field - the ```password``` and
+change the ```connection_string``` with a masked version. 
+
+To get the full connection string we add a utility function 
+```go
+package ent
+
+import "strings"
+
+// FullConnectionString returns the connection string with an unmasked password.
+func (u *User) FullConnectionString() string {
+	return strings.ReplaceAll(u.ConnectionString, "****", u.Password)
+}
+```
+
+BTW, you can see that it's OK to add files to the ent directory (not only in schema folder) since the ent generate command only appends/changes ent
+generated files.
+
+
+### Special Validators:
+
+Ent provides [validators](https://entgo.io/docs/schema-fields#validators) for fields, yet sometimes the logic can contain
+dependencies that are more applicative and require a broader context.
+For example, in our use case, we would like to make sure that dogs names don’t start with the first 2 letters of the owners name 
+(Best practice when you don't want your dog to run in circles every time someone calls your name).
+
+We start with a simple test
+```go
+func TestDogNameValidationHook(t *testing.T) {
+	ctx := context.Background()
+	c := enttest.Open(t, dialect.SQLite,
+		"file:TestSchemaConfHooks?mode=memory&cache=shared&_fk=1",
+	)
+	u := c.User.Create().SetName("Yoni").
+		SetConnectionString("mysql://root:pass@localhost:3306)").
+		SaveX(ctx)
+	_, err := c.Dog.Create().SetName("Yolo").SetOwner(u).Save(ctx)
+	require.Error(t, err)
+}
+```
+
+In this case, you can see that since "Yoni" and "Yolo" share the same 2 letters "Yo" the action fails.
+
+#### Recipe:
+
+First, register the hook:
+```go
+// Hooks of the Dog.
+func (Dog) Hooks() []ent.Hook {
+	return []ent.Hook{		
+		hook.If(validateName,
+			hook.HasFields(dog.FieldOwnerID),
+		),
+	}
+}
+```
+
+Since we want our hook to call every time we change the owner (to make sure the name is matched against), I am using an
+[edge field](https://entgo.io/docs/schema-edges#edge-field), this provides the hooks a trigger on when the edge's id changes.
+
+The hook:
+```go
+func validateName(next ent.Mutator) ent.Mutator {
+	return hook.DogFunc(func(ctx context.Context, m *gen.DogMutation) (ent.Value, error) {
+		owID, ok := m.OwnerID()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		owner, err := m.Client().User.Query().Where(user.ID(owID)).Only(ctx)
+		if err != nil {
+			return next.Mutate(ctx, m)
+		}
+		dn, ok := m.Name()
+		if !ok {
+			return next.Mutate(ctx, m)
+		}
+		if owner.Name[0:1] == dn[0:1] {
```

Comment:
You are talking about the first two letters, but you only check the very first one here :wink:

Reply:
👍